includes/actions/replace-product.php

```diff
@@ -0,0 +1,208 @@
+	 *
+	 * @var bool
+	 */
+	protected $allow_variable_products = true;
+
+	/**
+	 * Edit the item managed by this class on the subscription passed in the workflow's trigger
+	 *
+	 * @param WC_Subscription $subscription The subscription object.
+	 * @throws \Exception When there is an error.
+	 */
+	public function edit_subscription( $subscription ) {
+		$this->set_parameters();
+
+		if ( $this->old_product_id === $this->new_product_id ) {
+			throw new \Exception( __( 'Both the products are same', 'automatewoo-subscriptions' ) );
+		}
+
+		$old_product = wc_get_product( $this->old_product_id );
+		$new_product = wc_get_product( $this->new_product_id );
+
+		if ( ! $old_product ) {
+			throw new \Exception( __( 'Invalid product to be removed', 'automatewoo-subscriptions' ) . ' ' . $old_product );
+		}
+
+		if ( ! $new_product ) {
+			throw new \Exception( __( 'Invalid product to be added', 'automatewoo-subscriptions' ) . ' ' . $new_product );
+		}
+
+		if ( 'variable' === $new_product->get_type() ) {
+			/* translators: %s product name */
+			throw new \Exception( sprintf( __( '%s is a variable product parent and cannot be added.', 'automatewoo-subscriptions' ), $new_product->get_name() ) );
+		}
+
+		$qty = 0;
+
+		foreach ( $subscription->get_items() as $item ) {
+			$item_product_id = $item->get_product_id();
+
+			if ( $this->old_product_id !== $item_product_id ) {
+				continue;
+			}
+
+			$qty = $item->get_quantity();
+			$subscription->remove_item( $item->get_id() );
+			break;
+		}
+
+		if ( $qty > 0 ) {
+			$subscription->add_product( $new_product, $qty );
+
+			$this->reapply_coupons( $subscription );
+			$subscription->calculate_totals( true );
+		} else {
+			throw new \Exception( sprintf( __( 'Product to be replaced not found in the subscription', 'automatewoo-subscriptions' ) . ' ' . $old_product->get_name() ) );
+		}
+	}
+
+	/**
+	 * Reapply coupons.
+	 *
+	 * @param  WC_Subscription $subscription The subscription object.
+	 */
+	private function reapply_coupons( $subscription ) {
+		$coupons = $subscription->get_coupon_codes();
+		foreach ( $coupons as $coupon_code ) {
+			$subscription->remove_coupon( wc_format_coupon_code( $coupon_code ) );
+		}
+		foreach ( $coupons as $coupon_code ) {
+			$subscription->apply_coupon( wc_format_coupon_code( $coupon_code ) );
+		}
+	}
+
+	/**
+	 * Store data required to replace the product in a subscription.
+	 *
+	 * @return array
+	 */
+	private function get_data() {
+		return array(
+			'old_product_id' => $this->get_option( 'old_product_id' ),
+			'new_product_id' => $this->get_option( 'new_product_id' ),
+		);
+	}
+
+	/**
+	 * Recalculate a subscription's totals.
+	 *
+	 * @param \WC_Subscription $subscription The subscription object.
+	 */
+	protected function recalculate_subscription_totals( $subscription ) {
+		if ( is_callable( array( $subscription, 'recalculate_coupons' ) ) ) {
+			$subscription->recalculate_coupons();
+		} else {
+			$subscription->calculate_totals();
+		}
+	}
```

Comment:
```suggestion
	 * Recalculate a subscription's totals.
	 *
	 * @param \WC_Subscription $subscription The subscription object.
	 */
	protected function recalculate_subscription_totals( $subscription ) {
		if ( is_callable( array( $subscription, 'recalculate_coupons' ) ) ) {
			$subscription->recalculate_coupons();
		} else {
			$subscription->calculate_totals();
		}
	}
```

@abhishek-pokhriyal  We can remove this as I don't see we are using it anywhere. Any thoughts on this?

Reply:
@kuldeep3 👍